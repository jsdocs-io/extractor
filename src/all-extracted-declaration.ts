import type {
  ExtractedClass,
  ExtractedClassConstructor,
  ExtractedClassMethod,
  ExtractedClassProperty,
} from "./extract-class";
import type { ExtractedEnum, ExtractedEnumMember } from "./extract-enum";
import type { ExtractedFunction } from "./extract-function";
import type {
  ExtractedInterface,
  ExtractedInterfaceCallSignature,
  ExtractedInterfaceConstructSignature,
  ExtractedInterfaceGetAccessor,
  ExtractedInterfaceIndexSignature,
  ExtractedInterfaceMethod,
  ExtractedInterfaceProperty,
  ExtractedInterfaceSetAccessor,
} from "./extract-interface";
import type { ExtractedNamespace } from "./extract-namespace";
import type { ExtractedTypeAlias } from "./extract-type-alias";
import type { ExtractedVariable } from "./extract-variable";

/**
`AllExtractedDeclaration` is the union of all possible declarations
that can be extracted, with some being found only in other declarations
(e.g., class method declarations are found only in a class declaration).
*/
export type AllExtractedDeclaration =
  | ExtractedVariable
  | ExtractedFunction
  | ExtractedClass
  | ExtractedClassConstructor
  | ExtractedClassProperty
  | ExtractedClassMethod
  | ExtractedInterface
  | ExtractedInterfaceProperty
  | ExtractedInterfaceMethod
  | ExtractedInterfaceConstructSignature
  | ExtractedInterfaceCallSignature
  | ExtractedInterfaceIndexSignature
  | ExtractedInterfaceGetAccessor
  | ExtractedInterfaceSetAccessor
  | ExtractedEnum
  | ExtractedEnumMember
  | ExtractedTypeAlias
  | ExtractedNamespace;

/**
`AllExtractedDeclarationKind` is the union of all discriminators
used to detect the kind of declaration.
*/
export type AllExtractedDeclarationKind = AllExtractedDeclaration["kind"];
