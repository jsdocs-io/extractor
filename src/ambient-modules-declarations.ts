import { ModuleDeclaration, Node, type Project } from "ts-morph";
import { isHidden } from "./is-hidden";
import { sourceFilePath } from "./source-file-path";

export type AmbientModulesDeclarationsReturn = {
  containerName: string;
  exportName: string;
  declaration: ModuleDeclaration;
}[];

export const ambientModulesDeclarations = (
  containerName: string,
  project: Project,
  pkgName?: string,
): AmbientModulesDeclarationsReturn => {
  const ambientModulesDeclarations = [];
  for (const symbol of project.getAmbientModules()) {
    for (const declaration of symbol.getDeclarations()) {
      if (isHidden(declaration) || !Node.isModuleDeclaration(declaration)) {
        continue;
      }
      if (pkgName && !sourceFilePath(declaration).startsWith(`/${pkgName}`)) {
        // Ignore ambient modules that are not from the analyzed package.
        continue;
      }
      const exportName = declaration.getName();
      // TODO:
      // Remove surrounding quotes and eventual spaces
      // const exportID = exportName
      //   .replace(/"|'/g, "")
      //   .replace(/\s/g, "_")
      //   .trim();
      ambientModulesDeclarations.push({
        containerName,
        exportName,
        declaration,
      });
    }
  }
  return ambientModulesDeclarations;
};
