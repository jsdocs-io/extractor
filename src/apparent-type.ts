import { TypeFormatFlags, ts, type Node } from "ts-morph";

export function apparentType(node: Node): string {
	// See:
	// https://github.com/dsherret/ts-morph/issues/453#issuecomment-427405736
	// https://github.com/dsherret/ts-morph/issues/453#issuecomment-667578386
	return node
		.getType()
		.getApparentType()
		.getText(
			node,
			ts.TypeFormatFlags.NoTruncation | TypeFormatFlags.UseAliasDefinedOutsideCurrentScope,
		)
		.replace(/^Number$/, "number")
		.replace(/^Boolean$/, "boolean")
		.replace(/^String$/, "string");
}
