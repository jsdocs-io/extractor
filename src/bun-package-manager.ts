import { Effect } from "effect";
import { execa } from "execa";
import { InstallPackageError, PackageManager } from "./package-manager";

/** @internal */
export const bunPackageManager = (bunPath = "bun") =>
	PackageManager.of({
		installPackage: ({ pkg, cwd }) =>
			Effect.gen(function* (_) {
				// Run `bun add <pkg> --verbose`.
				// See https://bun.sh/docs/cli/add.
				const { stdout } = yield* _(
					Effect.tryPromise({
						try: () => execa(bunPath, ["add", pkg, "--verbose"], { cwd }),
						catch: (e) => new InstallPackageError({ cause: e }),
					}),
				);

				// With verbose output on, bun prints one line per installed package
				// (e.g., "foo@1.0.0"), including all installed dependencies.
				// These lines are between the two delimiting lines found here:
				// https://github.com/oven-sh/bun/blob/972a7b7080bd3066b54dcb43e9c91c5dfa26a69c/src/install/lockfile.zig#L5369-L5370.
				const lines = stdout.split("\n");
				const beginHash = lines.findIndex((line) => line.startsWith("-- BEGIN SHA512/256"));
				const endHash = lines.findIndex((line) => line.startsWith("-- END HASH"));
				const installedPackages = lines.slice(beginHash + 1, endHash);
				return installedPackages;
			}),
	});
