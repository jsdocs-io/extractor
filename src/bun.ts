import { execa } from "execa";

/** `Bun` is a wrapper for the Bun package manager. */
export class Bun {
	#bunCmd: string;

	constructor(bunCmd = "bun") {
		this.#bunCmd = bunCmd.trim();
	}

	/**
	`add` installs the given package in the given directory and
	returns a list of all installed packages (e.g., `["foo@1.0.0"]`).

	@see {@link https://bun.sh/docs/cli/add}
	*/
	async add(pkg: string, cwd: string): Promise<string[]> {
		// Run `bun add <pkg> --verbose` in the `cwd`.
		const { stdout } = await execa(this.#bunCmd, ["add", pkg, "--verbose"], { cwd });

		// With verbose output on, bun prints one line per installed package
		// (e.g., "foo@1.0.0"), including all installed dependencies.
		// These lines are between the two delimiting lines found here:
		// https://github.com/oven-sh/bun/blob/972a7b7080bd3066b54dcb43e9c91c5dfa26a69c/src/install/lockfile.zig#L5369-L5370.
		const lines = stdout.split("\n");
		const beginHash = lines.findIndex((line) => line.startsWith("-- BEGIN SHA512/256"));
		const endHash = lines.findIndex((line) => line.startsWith("-- END HASH"));
		const installedPackages = lines.slice(beginHash + 1, endHash);
		return installedPackages;
	}
}
