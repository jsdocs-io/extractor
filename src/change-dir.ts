import { Result } from "neverthrow";
import { OsError } from "./errors";

export const changeDir = Result.fromThrowable(
  process.chdir,
  (e) => new OsError("failed to change directory", { cause: e }),
);
