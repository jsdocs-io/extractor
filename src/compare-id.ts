export type WithId = {
  id: string;
};

export const compareId = (a: WithId, b: WithId): number =>
  a.id.localeCompare(b.id);
