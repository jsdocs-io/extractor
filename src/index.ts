export {
  extractDeclarations,
  type ExtractDeclarationsOptions,
  type ExtractedDeclaration,
  type ExtractedDeclarationKind,
} from "./extract-declarations";
