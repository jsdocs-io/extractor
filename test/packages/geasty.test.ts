import { expect, test } from "vitest";
import { extractPackageApi } from "../../src";

test("geasty@0.0.6", async () => {
	expect(
		await extractPackageApi({
			pkg: "geasty@0.0.6",
		}),
	).toMatchSnapshot({
		analyzedAt: expect.any(String),
		analyzedIn: expect.any(Number),
		packages: expect.any(Array),
	});
});
