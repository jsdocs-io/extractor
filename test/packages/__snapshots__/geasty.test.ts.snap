// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`geasty@0.0.6 1`] = `
{
  "analyzedAt": Any<String>,
  "analyzedIn": Any<Number>,
  "declarations": [
    {
      "constructors": [
        {
          "docs": [
            "/**
   * @param options Geasty options
   * @param options.access_token Fine-grained personal access tokens
   *
   * @example
   * \`\`\`ts
   * const geasty = new Geast({
   *   access_token: 'your_access_token_created_by_github'
   * })
   * \`\`\`
   */",
          ],
          "file": "/geasty/dist/index.d.mts",
          "id": "+class.default.constructor",
          "kind": "class-constructor",
          "line": 229,
          "name": "constructor",
          "signature": "constructor(options?: GeastyOptions);",
        },
      ],
      "docs": [],
      "file": "/geasty/dist/index.d.mts",
      "id": "+class.default",
      "kind": "class",
      "line": 205,
      "methods": [
        {
          "docs": [
            "/**
   * Allows you to add a new gist with one or more files.
   *
   * @param options
   * @param options.description Description of the gist
   * @param options.public Flag indicating whether the gist is public
   * @param options.files Names and content for the files that make up the gist
   * @returns The created gist
   *
   * @example
   * \`\`\`ts
   * creaetAGist({
   *   description: 'Example Gist',
   *   public: true,
   *   files: {'test.txt': {content: 'Hello World!'}},
   * })
   * \`\`\`
   */",
          ],
          "file": "/geasty/dist/index.d.mts",
          "id": "+class.default.+method.createAGist",
          "kind": "class-method",
          "line": 248,
          "name": "createAGist",
          "signature": "createAGist: (options: CreateAGistOptions) => Promise<Gist>;",
        },
        {
          "docs": [
            "/**
   * Delete a gist.
   * The fine-grained token must have the following permission set:
   *   - "Gists" user permissions (write)
   *
   * @param gistId The unique identifier of the gist.
   *
   * @example
   * \`\`\`ts
   * deleteAGist('gist_id')
   * \`\`\`
   */",
          ],
          "file": "/geasty/dist/index.d.mts",
          "id": "+class.default.+method.deleteAGist",
          "kind": "class-method",
          "line": 261,
          "name": "deleteAGist",
          "signature": "deleteAGist: (gistId: string) => Promise<void>;",
        },
        {
          "docs": [
            "/**
   * Fork a gist.
   * The fine-grained token must have the following permission set:
   *   - "Gists" user permissions (write)
   *
   * @param gistId The unique identifier of the gist.
   *
   * @example
   * \`\`\`ts
   * forkAGist('gist_id')
   * \`\`\`
   */",
          ],
          "file": "/geasty/dist/index.d.mts",
          "id": "+class.default.+method.forkAGist",
          "kind": "class-method",
          "line": 474,
          "name": "forkAGist",
          "signature": "forkAGist: (gistId: string) => Promise<void>;",
        },
        {
          "docs": [
            "/**
   * Gets a specified gist.
   *
   * @param gistId The unique identifier of the gist.
   * @returns Gist that matches the gist ID
   *
   * @example
   * \`\`\`ts
   * getAGist('gist_id')
   * \`\`\`
   */",
          ],
          "file": "/geasty/dist/index.d.mts",
          "id": "+class.default.+method.getAGist",
          "kind": "class-method",
          "line": 313,
          "name": "getAGist",
          "signature": "getAGist: (gistId: string) => Promise<Gist>;",
        },
        {
          "docs": [
            "/**
   * Gets a specified gist revision.
   *
   * @param gistId The unique identifier of the gist
   * @param sha The sha of the gist revision
   * @returns Gist revision that matches the gist ID and sha
   *
   * @example
   * \`\`\`ts
   * getAGistRevision('gist_id', 'sha')
   * \`\`\`
   */",
          ],
          "file": "/geasty/dist/index.d.mts",
          "id": "+class.default.+method.getAGistRevision",
          "kind": "class-method",
          "line": 347,
          "name": "getAGistRevision",
          "signature": "getAGistRevision: (gistId: string, sha: string) => Promise<Gist[]>;",
        },
        {
          "docs": [
            "/**
   * Lists the authenticated user's gists or if called anonymously, this returns all public gists.
   *
   * @param options
   * @param options.since Only show results that were last updated after the given time. This is a timestamp in ISO 8601 format: YYYY-MM-DDTHH:MM:SSZ.
   * @param options.page The page number of the results to fetch.
   * @param options.per_page The number of results per page (max 100).
   * @returns List of gists
   *
   * @example
   * \`\`\`ts
   * getAllGists({
   *   since: '2023-01-01T00:00:00Z',
   *   page: 1,
   *   per_page: 10,
   * })
   * \`\`\`
   */",
          ],
          "file": "/geasty/dist/index.d.mts",
          "id": "+class.default.+method.getAllGists",
          "kind": "class-method",
          "line": 301,
          "name": "getAllGists",
          "signature": "getAllGists: (options?: GetGistsOptions) => Promise<Gist[]>;",
        },
        {
          "docs": [
            "/**
   * List gist commits.
   *
   * @param options
   * @param options.gistId The unique identifier of the gist.
   * @param options.page The page number of the results to fetch.
   * @param options.per_page The number of results per page (max 100).
   * @returns List of gist commits
   *
   * @example
   * \`\`\`ts
   * getGistCommits({
   *   gistId: 'gist_id',
   *   page: 1,
   *   per_page: 10,
   * })
   * \`\`\`
   */",
          ],
          "file": "/geasty/dist/index.d.mts",
          "id": "+class.default.+method.getGistCommits",
          "kind": "class-method",
          "line": 423,
          "name": "getGistCommits",
          "signature": "getGistCommits: (
  options: GetGistForksOrCommitsOptions,
) => Promise<GistCommit[]>;",
        },
        {
          "docs": [
            "/**
   * List gist forks.
   *
   * @param options
   * @param options.gistId The unique identifier of the gist.
   * @param options.page The page number of the results to fetch.
   * @param options.per_page The number of results per page (max 100).
   * @returns List of gist forks
   *
   * @example
   * \`\`\`ts
   * getGistForks({
   *   gistId: 'gist_id',
   *   page: 1,
   *   per_page: 10,
   * })
   * \`\`\`
   */",
          ],
          "file": "/geasty/dist/index.d.mts",
          "id": "+class.default.+method.getGistForks",
          "kind": "class-method",
          "line": 404,
          "name": "getGistForks",
          "signature": "getGistForks: (options: GetGistForksOrCommitsOptions) => Promise<Gist[]>;",
        },
        {
          "docs": [
            "/**
   * Lists public gists for the specified user.
   *
   * @param options
   * @param options.username The handle for the GitHub user account
   * @param options.since Only show results that were last updated after the given time. This is a timestamp in ISO 8601 format: YYYY-MM-DDTHH:MM:SSZ.
   * @param options.page The page number of the results to fetch.
   * @param options.per_page The number of results per page (max 100).
   * @returns List of public gists for the specified user
   *
   * @example
   * \`\`\`ts
   * getGistsForUser({
   *   username: 'github_username',
   *   since: '2023-01-01T00:00:00Z',
   *   page: 1,
   *   per_page: 10,
   * })
   * \`\`\`
   */",
          ],
          "file": "/geasty/dist/index.d.mts",
          "id": "+class.default.+method.getGistsForUser",
          "kind": "class-method",
          "line": 334,
          "name": "getGistsForUser",
          "signature": "getGistsForUser: (options: GetGistForUserOptions) => Promise<Gist[]>;",
        },
        {
          "docs": [
            "/**
   * List public gists sorted by most recently updated to least recently updated.
   *
   * @param options
   * @param options.since Only show results that were last updated after the given time. This is a timestamp in ISO 8601 format: YYYY-MM-DDTHH:MM:SSZ.
   * @param options.page The page number of the results to fetch.
   * @param options.per_page The number of results per page (max 100).
   * @returns List of gists
   *
   * @example
   * \`\`\`ts
   * getPublicGists({
   *   since: '2023-01-01T00:00:00Z',
   *   page: 1,
   *   per_page: 10,
   * })
   * \`\`\`
   */",
          ],
          "file": "/geasty/dist/index.d.mts",
          "id": "+class.default.+method.getPublicGists",
          "kind": "class-method",
          "line": 366,
          "name": "getPublicGists",
          "signature": "getPublicGists: (options?: GetGistsOptions) => Promise<Gist[]>;",
        },
        {
          "docs": [
            "/**
   * Get the raw content of a gist file.
   *
   * @param options
   * @param options.username The handle for the GitHub user account
   * @param options.gistId The unique identifier of the gist
   * @param options.filename The name of the file within the gist. If not provided, the raw content of the first file in the gist will be returned.
   * @returns Raw content of the gist file
   *
   * @example
   * \`\`\`ts
   * getRawGistFileContent({
   *   username: 'github_username',
   *   gistId: 'gist_id',
   *   filename: 'file_name.txt',
   * })
   * \`\`\`
   */",
          ],
          "file": "/geasty/dist/index.d.mts",
          "id": "+class.default.+method.getRawGistFileContent",
          "kind": "class-method",
          "line": 493,
          "name": "getRawGistFileContent",
          "signature": "getRawGistFileContent: (options: {
  username: string;
  gistId: string;
  filename?: string;
}) => Promise<any>;",
        },
        {
          "docs": [
            "/**
   * List the authenticated user's starred gists.
   *
   * @param options
   * @param options.since Only show results that were last updated after the given time. This is a timestamp in ISO 8601 format: YYYY-MM-DDTHH:MM:SSZ.
   * @param options.page The page number of the results to fetch.
   * @param options.per_page The number of results per page (max 100).
   * @returns List of starred gists
   *
   * @example
   * \`\`\`ts
   * getStarredGists({
   *   since: '2023-01-01T00:00:00Z',
   *   page: 1,
   *   per_page: 10,
   * })
   * \`\`\`
   */",
          ],
          "file": "/geasty/dist/index.d.mts",
          "id": "+class.default.+method.getStarredGists",
          "kind": "class-method",
          "line": 385,
          "name": "getStarredGists",
          "signature": "getStarredGists: (options?: GetGistsOptions) => Promise<Gist[]>;",
        },
        {
          "docs": [
            "/**
   * Check if access token is provided.
   *
   * @returns Boolean indicating whether access token is provided
   */",
          ],
          "file": "/geasty/dist/index.d.mts",
          "id": "+class.default.+method.hasAccessToken",
          "kind": "class-method",
          "line": 512,
          "name": "hasAccessToken",
          "signature": "hasAccessToken: () => boolean;",
        },
        {
          "docs": [
            "/**
   * Check if a gist is starred.
   *
   * @param gistId The unique identifier of the gist.
   * @returns Boolean indicating whether the gist is starred
   *
   * @example
   * \`\`\`ts
   * isGistStarred('gist_id')
   * \`\`\`
   */",
          ],
          "file": "/geasty/dist/index.d.mts",
          "id": "+class.default.+method.isGistStarred",
          "kind": "class-method",
          "line": 435,
          "name": "isGistStarred",
          "signature": "isGistStarred: (gistId: string) => Promise<boolean>;",
        },
        {
          "docs": [
            "/**
   * Star a gist.
   * The fine-grained token must have the following permission set:
   *   - "Gists" user permissions (write)
   *
   * @param gistId The unique identifier of the gist.
   *
   * @example
   * \`\`\`ts
   * starAGist('gist_id')
   * \`\`\`
   */",
          ],
          "file": "/geasty/dist/index.d.mts",
          "id": "+class.default.+method.starAGist",
          "kind": "class-method",
          "line": 448,
          "name": "starAGist",
          "signature": "starAGist: (gistId: string) => Promise<void>;",
        },
        {
          "docs": [
            "/**
   * Unstar a gist.
   * The fine-grained token must have the following permission set:
   *   - "Gists" user permissions (write)
   *
   * @param gistId The unique identifier of the gist.
   *
   * @example
   * \`\`\`ts
   * unstarAGist('gist_id')
   * \`\`\`
   */",
          ],
          "file": "/geasty/dist/index.d.mts",
          "id": "+class.default.+method.unstarAGist",
          "kind": "class-method",
          "line": 461,
          "name": "unstarAGist",
          "signature": "unstarAGist: (gistId: string) => Promise<void>;",
        },
        {
          "docs": [
            "/**
   * Allows you to update a gist's description and to update, delete, or rename gist files. Files from the previous version of the gist that aren't explicitly changed during an edit are unchanged.
   * The fine-grained token must have the following permission set:
   *   - "Gists" user permissions (write)
   *
   * @param options
   * @param options.gistId The unique identifier of the gist.
   * @param options.description The description of the gist.
   * @param options.files The gist files to be updated, renamed, or deleted.
   * @returns Updated gist
   *
   * @example
   * \`\`\`ts
   * updateAGist({
   *   gistId: 'gist_id',
   *   description: 'Updated Description',
   *   files: {'test.txt': {content: 'Hello Geasty!'}},
   * })
   * \`\`\`
   */",
          ],
          "file": "/geasty/dist/index.d.mts",
          "id": "+class.default.+method.updateAGist",
          "kind": "class-method",
          "line": 282,
          "name": "updateAGist",
          "signature": "updateAGist: (
  options:
    | WithRequired<UpdateAGistOptions, "description">
    | WithRequired<UpdateAGistOptions, "files">,
) => Promise<Gist>;",
        },
      ],
      "name": "default",
      "properties": [],
      "signature": "class Geasty {}",
    },
  ],
  "name": "geasty",
  "overview": undefined,
  "packages": Any<Array>,
  "subpath": ".",
  "types": "./dist/index.d.mts",
  "version": "0.0.6",
}
`;
